crates/services/p2p/src/cached_view.rs

```diff
@@ -41,41 +41,64 @@ impl CachedView {
         }
     }
 
-    pub(crate) fn get_sealed_headers<V>(
+    fn get_from_cache_or_db<V, T, F>(
         &self,
+        cache: &DashMap<u32, T>,
         view: &V,
-        block_height_range: Range<u32>,
-    ) -> StorageResult<Option<Vec<SealedBlockHeader>>>
+        range: Range<u32>,
+        fetch_fn: F,
+    ) -> StorageResult<Option<Vec<T>>>
     where
         V: P2pDb,
+        T: Clone,
+        F: Fn(&V, Range<u32>) -> StorageResult<Option<Vec<T>>>,
     {
-        let mut headers = Vec::new();
-        let mut missing_ranges = block_height_range.clone();
-        for block_height in block_height_range.clone() {
-            if let Some(header) = self.sealed_block_headers.get(&block_height) {
-                headers.push(header.clone());
+        let mut items = Vec::new();
+        let mut missing_start = None;
+
+        for height in range.clone() {
+            if let Some(item) = cache.get(&height) {
+                items.push(item.clone());
             } else {
-                // for the first block not in the cache, start a new range
-                missing_ranges.start = block_height;
+                missing_start = Some(height);
                 break;
             }
         }
 
-        if missing_ranges.is_empty() {
+        if missing_start.is_none() {
```

Comment:
We can remove the sequential coupling (i.e. the safety of the `unwrap` below depending on this check) here by using a `let else` statement.

```suggestion
        let Some(missing_start) = missing_start else {
```

Reply:
addressed in [f16cb2f](https://github.com/FuelLabs/fuel-core/pull/2352/commits/f16cb2f0435fa3b478ae615ea4ba908acdabbf2e)